packages/swarm-components/src/Dropdown.jsx

```diff
@@ -0,0 +1,462 @@
+	);
+};
+
+Menu.propTypes = {
+	children: oneOfType([func, node]),
+};
+
+////////////////////////////////////////////////////////////////////////
+let MenuButton = React.forwardRef(
+	({ onClick, onKeyDown, onMouseDown, ...props }, ref) => {
+		const { refs, state, setState } = useContext(MenuContext);
+		const buttonRef = useRef(null);
+		const buttonRect = useRect(buttonRef, state.isOpen);
+
+		useEffect(() => {
+			if (!state.isOpen) {
+				buttonRef.current.focus();
+			}
+		}, [state.isOpen]);
+
+		useEffect(() => {
+			setState({ ...state, buttonRect });
+		}, [state.isOpen, window.scrollX, window.scrollY]);
+
+		return (
+			<ForwardedButton
+				id={state.buttonId}
+				aria-haspopup="menu"
+				aria-expanded={state.isOpen}
+				data-reach-menu-button
+				type="button"
+				ref={buttonRef}
+				onMouseDown={wrapEvent(onMouseDown, () => {
+					if (state.isOpen) {
+						setState({ ...state, closingWithClick: true });
+					}
+				})}
+				onClick={wrapEvent(onClick, () => {
+					if (state.isOpen) {
+						setState({ ...state, ...close() });
+					} else {
+						setState({ ...state, ...openAtFirstItem() });
+					}
+				})}
+				onKeyDown={wrapEvent(onKeyDown, event => {
+					if (event.key === 'ArrowDown') {
+						event.preventDefault(); // prevent scroll
+						setState({ ...state, ...openAtFirstItem() });
+					} else if (event.key === 'ArrowUp') {
+						event.preventDefault(); // prevent scroll
+						setState({ ...state, ...openAtFirstItem() });
+					}
+				})}
+				{...props}
+			/>
+		);
+	}
+);
+
+MenuButton.propTypes = {
+	onClick: func,
+	onKeyDown: func,
+	children: node,
+};
+
+let MenuItem = React.forwardRef(
+	(
+		{
+			onSelect,
+			onClick,
+			role = 'menuitem',
+			index,
+			onKeyDown,
+			onMouseMove,
+			onMouseLeave,
+			setState: propSetState,
+			state: propState,
+			_ref,
+			...rest
+		},
+		ref
+	) => {
+		const { state, setState } = useContext(MenuContext);
+		let isSelected = state && index === state.selectionIndex;
+		let select = () => {
+			onSelect();
+			setState({ ...state, ...close() });
+		};
+
+		const itemRef = useRef(null);
+
+		// assignRef(ref, node);
+		// assignRef(_ref, node);
+
+		useEffect(() => {
+			if (itemRef.current && state.selectionIndex === index) {
+				itemRef.current.focus();
+			}
+		}, [state.selectionIndex]);
+		return (
+			<div
+				{...rest}
+				ref={itemRef}
+				data-reach-menu-item={role === 'menuitem' ? true : undefined}
+				role={role}
+				tabIndex="-1"
+				data-selected={role === 'menuitem' && isSelected ? true : undefined}
+				onClick={wrapEvent(onClick, event => {
+					select();
+				})}
+				onKeyDown={wrapEvent(onKeyDown, event => {
+					if (event.key === 'Enter') {
+						// prevent the button from being "clicked" by
+						// this "Enter" keydown
+						event.preventDefault();
+						select();
+					}
+				})}
+				onMouseMove={wrapEvent(onMouseMove, event => {
+					if (!isSelected) {
+						setState({ ...state, ...selectItemAtIndex(index) });
```

Comment:
you don't need to send the full `state` object to `setState` - just the part that needs to be set, e.g. `setState(selectItemAtIndex(index));` here

Reply:
We do here since this `setState` is an updater of `useState()` and updates slightly differently than the `setState` we're used to by replacing the state with what is passed rather than updating the corresponding key in state